File: instagram-handler/services/proxy_manager.py
import os
import random
import time
from typing import Any

import requests

CACHE_FILE = os.path.join(os.path.dirname(os.path.dirname(__file__)), "data", "proxy_cache.json")
CACHE_TTL = 300  # 5 minutes
_index = 0


def get_proxy_url() -> str:
    return os.getenv(
        "WEBSHARE_PROXY_URL",
        "",
    ).strip()


def _parse_line(line: str) -> str:
    line = line.strip()
    if not line or line.startswith("#"):
        return ""
    parts = line.split(":")
    if len(parts) < 4:
        return ""
    host, port, user = parts[0], parts[1], parts[2]
    password = ":".join(parts[3:])
    return f"http://{user}:{password}@{host}:{port}"


def fetch_proxies(force: bool = False) -> list[str]:
    global _index
    url = get_proxy_url()
    if not url:
        return []

    if not force and os.path.exists(CACHE_FILE):
        try:
            import json

            with open(CACHE_FILE, encoding="utf-8") as f:
                cache = json.load(f)
            if time.time() - cache.get("fetched_at", 0) < CACHE_TTL:
                return cache.get("proxies", [])
        except (json.JSONDecodeError, OSError):
            pass

    resp = requests.get(url, timeout=30, headers={"User-Agent": "IG-Handler/1.0"})
    resp.raise_for_status()
    proxies = [_parse_line(line) for line in resp.text.splitlines()]
    proxies = [p for p in proxies if p]

    os.makedirs(os.path.dirname(CACHE_FILE), exist_ok=True)
    import json

    with open(CACHE_FILE, "w", encoding="utf-8") as f:
        json.dump({"fetched_at": time.time(), "proxies": proxies}, f)

    _index = 0
    return proxies


def get_next_proxy() -> str:
    global _index
    proxies = fetch_proxies()
    if not proxies:
        return ""
    proxy = proxies[_index % len(proxies)]
    _index += 1
    return proxy


def get_random_proxy() -> str:
    proxies = fetch_proxies()
    return random.choice(proxies) if proxies else ""


def resolve_proxy(explicit: str = "", auto: bool = True) -> str:
    if explicit and explicit.strip():
        return explicit.strip()
    if auto and get_proxy_url():
        return get_next_proxy()
    return ""


def get_stats() -> dict[str, Any]:
    proxies = fetch_proxies()
    return {
        "enabled": bool(get_proxy_url()),
        "total_proxies": len(proxies),
        "cache_ttl_seconds": CACHE_TTL,
        "provider": "webshare",
    }

File: instagram-handler-php/src/ProxyManager.php
<?php

declare(strict_types=1);

final class ProxyManager
{
    private const CACHE_TTL = 300;
    private static int $index = 0;

    public static function getProxyUrl(): string
    {
        return trim(getenv('WEBSHARE_PROXY_URL') ?: '');
    }

    public static function cacheFile(): string
    {
        return DATA_PATH . '/proxy_cache.json';
    }

    public static function parseLine(string $line): string
    {
        $line = trim($line);
        if ($line === '' || str_starts_with($line, '#')) {
            return '';
        }
        $parts = explode(':', $line);
        if (count($parts) < 4) {
            return '';
        }
        $host = $parts[0];
        $port = $parts[1];
        $user = $parts[2];
        $password = implode(':', array_slice($parts, 3));
        return "http://{$user}:{$password}@{$host}:{$port}";
    }

    public static function fetchProxies(bool $force = false): array
    {
        $url = self::getProxyUrl();
        if ($url === '') {
            return [];
        }

        $cacheFile = self::cacheFile();
        if (!$force && is_file($cacheFile)) {
            $cache = json_decode((string) file_get_contents($cacheFile), true);
            if (is_array($cache) && (time() - ($cache['fetched_at'] ?? 0)) < self::CACHE_TTL) {
                return $cache['proxies'] ?? [];
            }
        }

        $ch = curl_init($url);
        curl_setopt_array($ch, [
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT => 30,
            CURLOPT_HTTPHEADER => ['User-Agent: IG-Handler-PHP/1.0'],
        ]);
        $body = curl_exec($ch);
        if ($body === false) {
            throw new RuntimeException('Webshare fetch failed: ' . curl_error($ch));
        }
        curl_close($ch);

        $proxies = [];
        foreach (explode("\n", $body) as $line) {
            $parsed = self::parseLine($line);
            if ($parsed !== '') {
                $proxies[] = $parsed;
            }
        }

        if (!is_dir(DATA_PATH)) {
            mkdir(DATA_PATH, 0755, true);
        }
        file_put_contents($cacheFile, json_encode([
            'fetched_at' => time(),
            'proxies' => $proxies,
        ]));

        self::$index = 0;
        return $proxies;
    }

    public static function getNextProxy(): string
    {
        $proxies = self::fetchProxies();
        if (!$proxies) {
            return '';
        }
        $proxy = $proxies[self::$index % count($proxies)];
        self::$index++;
        return $proxy;
    }

    public static function getRandomProxy(): string
    {
        $proxies = self::fetchProxies();
        if (!$proxies) {
            return '';
        }
        return $proxies[array_rand($proxies)];
    }

    public static function resolveProxy(string $explicit = '', bool $auto = true): string
    {
        $explicit = trim($explicit);
        if ($explicit !== '') {
            return $explicit;
        }
        if ($auto && self::getProxyUrl() !== '') {
            return self::getNextProxy();
        }
        return '';
    }

    public static function getStats(): array
    {
        $proxies = self::fetchProxies();
        return [
            'enabled' => self::getProxyUrl() !== '',
            'total_proxies' => count($proxies),
            'cache_ttl_seconds' => self::CACHE_TTL,
            'provider' => 'webshare',
        ];
    }
}
